
1. telldir is completely broken.
2. After fixing it, a call to telldir right after a seekdir (without a
   readdir between the calls) reports a wrong value.

- Gernot Tenchio <g.tenchio@telco-tech.de>

--- dietlibc-cvs/lib/seekdir.c.telldir	2006-03-07 17:42:34.000000000 +0100
+++ dietlibc-cvs/lib/seekdir.c	2006-03-07 17:44:41.000000000 +0100
@@ -3,6 +3,8 @@
 #include <dirent.h>
 
 void seekdir(DIR *d,off_t offset) {
-  if (lseek(d->fd,offset,SEEK_SET) != (off_t)-1)
+  if (lseek(d->fd,offset,SEEK_SET) != (off_t)-1) {
     d->num=d->cur=0;
+    ((struct dirent *)(d->buf))->d_off = offset;
+  }
 }
--- dietlibc-cvs/lib/telldir.c.telldir	2006-03-07 17:47:23.000000000 +0100
+++ dietlibc-cvs/lib/telldir.c	2006-03-07 17:46:58.000000000 +0100
@@ -3,5 +3,8 @@
 #include <dirent.h>
 
 off_t telldir(DIR *d) {
-  return lseek(d->fd,0,SEEK_CUR)-d->num+d->cur;
+  off_t result = 0;
+  if (lseek(d->fd,0,SEEK_CUR))
+    result=((struct dirent*)(d->buf+d->cur))->d_off;
+  return result;
 }
